### PR TITLE
Add ECC tools

### LINKED ISSUE #829: Implement commit
Implement a tool to use the commit command.

### LINKED ISSUE #834: Implement ec_ephemeral
Implement a tool to use the ec_ephemeral command.

### LINKED ISSUE #835: Implement ecc_parameters
Implement a tool to use the ecc_parameters command.

### LINKED ISSUE #877: Implement zgen_2phase
Implement a tool to use the zgen_2phase command.

### LINKED ISSUE #2026: Implement TPM2_ECDH_Keygen

### LINKED ISSUE #2028: Implement TPM2_ECDH_ZGen

### PR DESCRIPTION
**Pull request by idesai**
Fixes #835 
Fixes #834
Fixes #829 
Fixes #2026 
Fixes #2028 
Fixes #877 

Signed-off-by: Imran Desai <imran.desai@intel.com>

### CODE REVIEW

File: tools/tpm2_commit.c [hunk 1]
```diff
@@ -0,0 +1,197 @@
...
+        return tool_rc_general_error;
+    }
+
+    TSS2_RC rval = Esys_Commit(ectx, ctx.signing_key.object.tr_handle,
+        signing_key_obj_session_handle, ESYS_TR_NONE, ESYS_TR_NONE, &ctx.P1,
+        &ctx.s2, &ctx.y2, &ctx.K, &ctx.L, &ctx.E, &ctx.counter);
+    if (rval != TSS2_RC_SUCCESS) {
+        LOG_PERR(Esys_Commit, rval);
+        return tool_rc_from_tpm(rval);
+    }
```
Comment (idesai) [inline tools/tpm2_commit.c:168]: move to tpm2.c

File: tools/tpm2_ecdhkeygen.c [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+        return rc;
+    }
+
+    TSS2_RC rval = Esys_ECDH_KeyGen(ectx, ctx.ecc_public_key.object.tr_handle,
+        ESYS_TR_NONE, ESYS_TR_NONE, ESYS_TR_NONE, &ctx.Z, &ctx.Q);
+    if (rval != TSS2_RC_SUCCESS) {
+        LOG_PERR(Esys_ECDH_KeyGen, rval);
+        return tool_rc_from_tpm(rval);
+    }
+
```
Comment (idesai) [inline tools/tpm2_ecdhkeygen.c:93]: move to tpm2.c

File: tools/tpm2_ecdhzgen.c [hunk 3]
```diff
@@ -0,0 +1,122 @@
...
+        LOG_ERR("Failed to load public input ECC point Q");
+        return tool_rc_general_error;
+    }
+
+    TSS2_RC rval = Esys_ECDH_ZGen(ectx, ctx.ecc_key.object.tr_handle,
+    ecc_key_obj_session_handle, ESYS_TR_NONE, ESYS_TR_NONE, &ctx.Q, &ctx.Z);
+    if (rval != TSS2_RC_SUCCESS) {
+        LOG_PERR(Esys_ECDH_ZGen, rval);
+        return tool_rc_from_tpm(rval);
+    }
```
Comment (idesai) [inline tools/tpm2_ecdhzgen.c:113]: move to tpm2.c

File: tools/tpm2_ecephermal.c [hunk 4]
```diff
@@ -0,0 +1,132 @@
...
+
+    TPM2B_ECC_POINT *Q;
+    uint16_t counter;
+    TSS2_RC rval = Esys_EC_Ephemeral(ectx, ESYS_TR_NONE, ESYS_TR_NONE,
+        ESYS_TR_NONE, ctx.curve_id, &Q, &counter);
+    if (rval != TSS2_RC_SUCCESS) {
+        LOG_PERR(Esys_EC_Ephemeral, rval);
+        return tool_rc_from_tpm(rval);
+    }
+
```
Comment (idesai) [inline tools/tpm2_ecephermal.c:115]: move to tpm2.c

File: tools/tpm2_geteccparameters.c [hunk 5]
```diff
@@ -0,0 +1,112 @@
...
+    }
+
+    TPMS_ALGORITHM_DETAIL_ECC *parameters;
+    TSS2_RC rval = Esys_ECC_Parameters(ectx, ESYS_TR_NONE, ESYS_TR_NONE,
+        ESYS_TR_NONE, ctx.curve_id, &parameters);
+    if (rval != TSS2_RC_SUCCESS) {
+        LOG_PERR(Esys_ECC_Parameters, rval);
+        return tool_rc_from_tpm(rval);
+    }
+
```
Comment (idesai) [inline tools/tpm2_geteccparameters.c:104]: move to tpm2.c

File: tools/tpm2_zgen2phase.c [hunk 6]
```diff
@@ -0,0 +1,189 @@
...
+        return tool_rc_general_error;
+    }
+
+    TSS2_RC rval = Esys_ZGen_2Phase(ectx, ctx.ecc_key.object.tr_handle,
+        ecc_key_obj_session_handle, ESYS_TR_NONE, ESYS_TR_NONE, &ctx.Q1,
+        &ctx.Q2, ctx.keyexchange_scheme, ctx.commit_counter, &ctx.Z1, &ctx.Z2);
+    if (rval != TSS2_RC_SUCCESS) {
+        LOG_PERR(Esys_ZGen_2Phase, rval);
+        return tool_rc_from_tpm(rval);
+    }
```
Comment (idesai) [inline tools/tpm2_zgen2phase.c:174]: move to tpm2.c

File: tools/tpm2_geteccparameters.c [hunk 7]
```diff
@@ -0,0 +1,112 @@
...
+static tpm_geteccparameters_ctx ctx = {
+    .curve_id = TPM2_ECC_NONE,
+};
+
+static bool on_option(char key, char *value) {
+
+    switch (key) {
+    case 'o':
+        ctx.ecc_parameters_path = value;
+        break;
```
Comment (idesai) [inline tools/tpm2_geteccparameters.c:23]: Print YAML out if the option "o" is not specified.

File: man/tpm2_commit.1.md [hunk 8]
```diff
@@ -0,0 +1,82 @@
...
+# SYNOPSIS
+
+**tpm2_commit** [*OPTIONS*]
+
+# DESCRIPTION
+
+**tpm2_commit**(1) - Performs the first part of an ECC anonymous signing
+operation. The TPM will perform the point multiplications on the provided points
+and return intermediate signing values. The signHandle parameter shall refer to
+an ECC key and the signing scheme must be anonymous (TPM_RC_SCHEME).
```
Comment (williamcroberts) [inline man/tpm2_commit.1.md:17]: i don't see a signHandle and I am also not getting what you mean by "signign scheme anonymous". Are you referring to the EC Key Objects scheme parameters being NULL? Below it looks like it needs an ECDAA scheme, is that what you meant?

File: man/tpm2_commit.1.md [hunk 9]
```diff
@@ -0,0 +1,82 @@
...
+
+    Specify the input data used to derive the x coordinate of the basepoint.
+
+  * **\--basepoint-y**=_FILE_:
+
+    Specify the y coordinate of the basepoint.
+
+  * **\--eccpoint-P**=_FILE_:
+
+    Specify a point on the curve used by sign handle.
```
Comment (williamcroberts) [inline man/tpm2_commit.1.md:31]: It looks like ARGUMENT, y and P can be omitted, what happens when they are omitted?

File: man/tpm2_ecdhzgen.1.md [hunk 10]
```diff
@@ -0,0 +1,63 @@
...
+    Context object pointing to ECC key.
+    Either a file or a handle number. See section "Context Object Format".
+
+  * **-p**, **\--key-auth**=_AUTH_:
+
+    The authorization value for the ECC key object.
+
+  * **-u**, **\--public**=_FILE_:
+
+    Output ECC point Q
```
Comment (williamcroberts) [inline man/tpm2_ecdhzgen.1.md:32]: Needs a period at the end of the sentence. Q is what you share with the other party during the ECDH exchange right?

File: man/tpm2_zgen2phase.1.md [hunk 11]
```diff
@@ -0,0 +1,83 @@
...
+
+    The authorization value for the ECC key object.
+
+  * **-s**, **\--scheme**=_ALGORITHM_:
+
+    The key exchange scheme. Optional. Valid options are ecdh or sm2.
+
+  * **-t**, **\--counter**=_NATURALNUMBER_:
+
+    The commit count to determine the key index to use.
```
Comment (williamcroberts) [inline man/tpm2_zgen2phase.1.md:37]: Why are the -t option here different from tpm2_ecephermal?

**Comment by idesai** [comment]
> Fix leaks reported by [ASAN](https://travis-ci.org/github/tpm2-software/tpm2-tools/jobs/688968714)

fixed

File: man/tpm2_ecdhzgen.1.md (hunk 10 above)
Comment (idesai) [reply man/tpm2_ecdhzgen.1.md:32]: correct

File: man/tpm2_zgen2phase.1.md (hunk 11 above)
Comment (idesai) [reply man/tpm2_zgen2phase.1.md:37]: In ephemeral, the counter value is incremented. The -t option there is to give the user an option to specify a descriptive file name that he can later use to invoke the specific key index at the time of running say zgen2phase. Note that the counter value increments with commit as well as ephemeral calls. While it is still a key index, the two can be told apart from a descriptive filename.

File: man/tpm2_commit.1.md (hunk 9 above)
Comment (idesai) [reply man/tpm2_commit.1.md:31]: These options give additional controls in determining the point on the curve. Defaults work just fine and that is why they are optional parameters. Should we go with the defaults for now?

File: man/tpm2_commit.1.md (hunk 8 above)
Comment (idesai) [reply man/tpm2_commit.1.md:17]: Yes it should be a ECDAA. Also the sign handle is the key and its auth specified with -p and -c. I will correct the text.

File: man/tpm2_commit.1.md (hunk 9 above)
Comment (idesai) [reply man/tpm2_commit.1.md:32]: Retaining these options for now.

File: tools/tpm2_geteccparameters.c (hunk 7 above)
Comment (idesai) [reply tools/tpm2_geteccparameters.c:24]: Address in a separate ticket and add the yaml support for all ECC tools.

**Pull request merged**